xarray/core/indexing.py

```diff
@@ -56,7 +56,8 @@ def canonicalize(indexer):
                 if indexer.ndim != 1:
                     raise ValueError('orthogonal array indexing only supports '
                                      '1d arrays')
-                if indexer.dtype.kind == 'b':
+                if (indexer.dtype.kind == 'b' or
+                        (indexer.dtype.kind == 'O' and len(indexer) == 0)):
```

Comment:
Can we fix this on `where` instead? I'm not sure it makes sense to allow `dtype=object` for all positional indexers, even for 0-d arrays.